### PR TITLE
Use posixpath for URLs

### PR DESCRIPTION
**Pull request by danlipsa**
This fixed 'spack install hdf5' produced the following error
 Error: FetchError: All fetchers failed for hdf5-1.14.1-2-yyfgptn

### CODE REVIEW

**Comment by danlipsa** [comment]
@johnwparent Please review

File: lib/spack/spack/util/web.py [hunk 1]
```diff
@@ -738,7 +739,8 @@ def find_versions_of_archive(
 
         # We'll be a bit more liberal and just look for the archive
         # part, not the full path.
-        url_regex = os.path.basename(url_regex)
+        # this is a URL so it is a posixpath even on Windows
+        url_regex = posixpath.basename(url_regex)
```
Comment (johnwparent) [inline lib/spack/spack/util/web.py]: We only support Python 3.7+ now, so lets use `Pathlib` over the `os.path` module. 
```suggestion
        url_regex = pathlib.PurePath(url_regex).name
```

**Comment by johnwparent** [comment]
Looks good, left one comment/suggestion that should be addressed, but otherwise LGTM, thanks for digging down on this

File: lib/spack/spack/util/web.py (hunk 1 above)
Comment (danlipsa) [reply lib/spack/spack/util/web.py]: Done. I use PurePosixPath though - the same idea as using posixpath instead of path.

**Review by johnwparent — DISMISSED** [review]

**Review by scheibelp — CHANGES_REQUESTED** [review]
Can you expand on the issue in the PR description? Which Spack commands produce the error? `spack install`? `spack checksum`? etc.

Also I think this PR should reenable or add tests to `lib/spack/spack/test/web.py` (which tests the function that was modified, but generally not on Windows).

**Comment by danlipsa** [comment]
@scheibelp I addressed all your comments. Can you take another look at this and merge if ready? Thanks!

**Review by scheibelp — APPROVED** [review]

**Pull request merged**